calico/felix/fetcd.py

```diff
@@ -60,6 +77,74 @@ def __init__(self, config):
         self.client = None
         self.my_config_dir = dir_for_per_host_config(self.config.HOSTNAME)
 
+        # Initialized at poll start time.
+        self.splitter = None
+        self.next_etcd_index = None
+
+        # Cache of known endpoints, used to resolve deletions of whole
+        # directory trees.
+        self.endpoint_ids_per_host = defaultdict(set)
+
+        self.handlers = {
+            "delete": self.force_resync,
+
+            "Ready": {
+                "set": self.on_ready,
+                "delete": self.force_resync,
+            },
+
+            "policy": {
+                "delete": self.force_resync,
+
+                "profile": {
+                    "delete": self.force_resync,
+
+                    "capture": ("profile_id",  {
+                        "delete": self.on_profile_delete,
+
+                        "tags": {
+                            "delete": self.on_tags_delete,
+                            "set": self.on_tags_set,
+                        },
+
+                        "rules": {
+                            "delete": self.on_rules_delete,
+                            "set": self.on_rules_set,
+                        }
+                    })
+                }
+            },
+
+            "host": {
+                "delete": self.force_resync,
+
+                "capture": ("hostname",  {
+                    "delete": self.on_host_delete,
+
+                    "workload": {
+                        "delete": self.on_host_delete,
+
+                        "capture": ("orchestrator",  {
+                            "delete": self.on_orch_delete,
+
+                            "capture": ("workload_id",  {
+                                "delete": self.on_workload_delete,
+
+                                "endpoint": {
+                                    "delete": self.on_workload_delete,
+
+                                    "capture": ("endpoint_id",  {
+                                        "delete": self.on_endpoint_delete,
+                                        "set": self.on_endpoint_set
+                                    })
+                                }
+                            })
+                        })
+                    }
+                })
+            }
+        }
+
```

Comment:
![](http://makeameme.org/media/created/what-language-is.jpg)

Is Felix written in Javascript now?

All joking aside, this is an excellent example of where a dictionary is a *terrible* design and interface choice, not least because populating this dict is such an obvious exercise in madness. Additionally, this data structure is presented with exactly zero comments explaining how it's supposed to be structured, which means I have got genuinely no idea what this is, what it's for, and why it looks the way it looks.

Pull this whole mess out (along with `handle_update`), split it into a class that allows handler registration, and then give a vaguely sensible API for handling that kind of registration. Something like:

```python
self.dispatcher = EventDispatcher()
self.dispatcher.register('/Ready', set=self.on_ready, delete=self.force_resync)
self.dispatcher.register('/policy', delete=self.force_resync)
self.dispatcher.register('/policy/profile', delete=self.force_resync)
self.dispatcher.register('/policy/profile/<profile_id>/', delete=self.on_profile_delete)
self.dispatcher.register(/policy/profile/<profile_id>/tags', set=self.on_tags_set, delete=self.on_tags_delete)

self.dispatcher.dispatch(...)
```

This is definitely going to be more work than this dictionary, but it's a small price to pay for code that is debuggable.